### PR TITLE
rewrite fillna3D using pandas

### PR DESCRIPTION
**Pull request by leiyangleon**
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have added an explanation of what your changes do and why you'd like us to include them.
- [ ] I have written new tests for your core changes, as applicable.
- [ ] I have successfully ran tests with your changes locally.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tools/RAiDER/interpolator.py [hunk 1]
```diff
@@ -91,21 +91,20 @@ def interpVector(vec, Nx):
...
-    def finder(z): return z.nonzero()[0]
-
-    try:
-        y[test_nan] = np.interp(finder(test_nan), finder(~test_nan), y[~test_nan])
-        newy = np.reshape(y, shape)
-        final = np.moveaxis(newy, -1, axis)
-        return final
-    except ValueError:
-        return array
+    import pandas as pd
```
Comment (jlmaurer) [inline tools/RAiDER/interpolator.py:94]: Move import to top of file

Comment (leiyangleon) [reply tools/RAiDER/interpolator.py]: did so just now

File: tools/RAiDER/interpolator.py [hunk 2]
```diff
@@ -91,21 +93,19 @@ def interpVector(vec, Nx):
...
+        array = temp_pd.to_numpy()
+    elif (axis == -1):
+        for yy in range(YY):
+            temp_pd = pd.DataFrame(data=array[:,yy,:], dtype=np.float32)
+            temp_pd = temp_pd.interpolate(method='linear',axis=1,limit_direction='both')
+            array[:,yy,:] = temp_pd.to_numpy()
+    else:
+        raise Exception("Axis out of the array shape")
+    
+    return array
```
Comment (jlmaurer) [inline tools/RAiDER/interpolator.py:111]: You can simplify this using np.moveaxis and reshape:

```suggestion
    narr = np.moveaxis(array,axis,-1)
    nars = narr.reshape((np.prod(narr.shape[:-1]),) +(narr.shape[-1],))
    dfd = pd.DataFrame(data=nars).interpolate(axis=1,limit_direction='both')
    out = dfd.values.reshape(array.shape)
    return np.moveaxis(out,-1,axis)
```

**Review by jlmaurer — CHANGES_REQUESTED** [review]
The unit test for fillna3D needs updated as well. Can you add the following to the test/test_interpolator.py module? (you can remove the current test_fillna3d function)
```
@pytest.fixture
def nanArr():
    array = np.random.randn(2,2,3)
    array[0,0,0] = np.nan
    array[0,0,1] = np.nan
    array[0,0,2] = np.nan
    array[1,0,0] = np.nan
    array[0,1,1] = np.nan
    array[1,1,2] = np.nan
    true_array = array.copy()
    true_array[0,0,0] = np.nan
    true_array[0,0,1] = np.nan
    true_array[0,0,2] = np.nan
    true_array[1,0,0] = true_array[1,0,1]
    true_array[0,1,1] = (true_array[0,1,0] + true_array[0,1,2])/2
    true_array[1,1,2] = true_array[1,1,1]
    return array, true_array

def test_fillna3D(nanArr):
    arr, tarr = nanArr
    assert np.allclose(fillna3D(arr), tarr, equal_nan=True)
```

File: tools/RAiDER/interpolator.py (hunk 2 above)
Comment (leiyangleon) [reply tools/RAiDER/interpolator.py]: Great. That is so concise! Really like the way you rewrote. I revised this function as well as the unit test per your suggestion. The CI check ran successfully for the interpolator unit test. Still failed at test_scenario_1 and test_scenario_2 for the reasons mentioned in my email. If you agree, I can re-create those true-value files, which will let the CI check go through and conclude this PR.

**Comment by jlmaurer** [comment]
Great @leiyangleon, looking good! Go ahead and update the test_scenario files so that those two unit tests pass and then we can merge this in. Thanks!

**Comment by leiyangleon** [comment]
> Great @leiyangleon, looking good! Go ahead and update the test_scenario files so that those two unit tests pass and then we can merge this in. Thanks!

@jlmaurer I updated the two unit test files and CI build was successful. However, the coverage test failed, which I have no idea what happened and how to track the errors by clicking on the link for details. Could you help?

**Comment by dbekaert** [comment]
@leiyangleon no worries for the coverage report. It fails when the coverage of the code decreases. It says that new code was added but the current test-suite does not not test all of the new code changes.

**Review by dbekaert — APPROVED** [review]

**Comment by jlmaurer** [comment]
@dbekaert this is quite confusing to me; I checked the coverage failure but the only changed file is 100% covered. Also it is not letting me merge the PR while the coverage test is failing.

**Comment by dbekaert** [comment]
You requested changes, which needs to be addressed before you can merge ;).
I overruled and merged it.

**Pull request merged**